skimage/segmentation/_join.py

```diff
@@ -116,26 +117,177 @@ def relabel_sequential(label_field, offset=1):
+    # Some logic to determine the output type:
+    #  - we don't want to return a smaller output type than the input type,
+    #  ie if we get uint32 as labels input, don't return a uint8 array.
+    #  - but, in some cases, using the input type could result in overflow. The
+    #  input type could be a signed integer (e.g. int32) but
+    #  `np.min_scalar_type` will always return an unsigned type. We check for
+    #  that by casting the largest output value to the input type. If it is
+    #  unchanged, we use the input type, else we use the unsigned minimum
+    #  required type
+    required_type = np.min_scalar_type(out_vals[-1])
+    if input_type.itemsize < required_type.itemsize:
         output_type = required_type
-    forward_map = np.zeros(max_label + 1, dtype=output_type)
-    forward_map[labels0] = new_labels0
-    inverse_map = np.zeros(new_max_label + 1, dtype=output_type)
-    inverse_map[offset:] = labels0
-    relabeled = forward_map[label_field]
-    return relabeled, forward_map, inverse_map
+    else:
+        if input_type.type(out_vals[-1]) == out_vals[-1]:
+            output_type = input_type
+        else:
+            output_type = required_type
+    out_array = np.empty(label_field.shape, dtype=output_type)
+    out_vals = out_vals.astype(output_type)
+    map_array(label_field, in_vals, out_vals, out=out_array)
+    fw_map = ArrayMap(in_vals, out_vals)
+    inv_map = ArrayMap(out_vals, in_vals)
+    return out_array, fw_map, inv_map
+
+
+def map_array(input_arr, input_vals, output_vals, out=None):
+    """Map values from input array from input_vals to output_vals.
+
+    Parameters
+    ----------
+    input_arr : array of int, shape (M[, N][, P][, ...])
+        The input label image.
+    input_vals : array of int, shape (N,)
+        The values to map from.
+    output_vals : array, shape (N,)
+        The values to map to.
+    out: array, same shape as `input_arr`
+        The output array. Will be created if not provided. It should
+        have the same dtype as `output_vals`.
+
+    Returns
+    -------
+    out : array, same shape as `input_arr`
+        The array of mapped values.
+    """
+
+    # We ravel the input array for simplicity of iteration in Cython:
+    orig_shape = input_arr.shape
+    # NumPy docs for `np.ravel()` says:
+    # "When a view is desired in as many cases as possible, 
+    # arr.reshape(-1) may be preferable."
+    input_arr = input_arr.reshape(-1)
+    if out is None:
+        out = np.empty_like(input_arr, dtype=output_vals.dtype)
+    else:
+        out_original = out
+        out = out_original.reshape(-1)
+        if out.strides[0] != min(out_original.strides):
+            # reshaping forced a copy
+            raise ValueError(
+                'If out array is provided, it should be either contiguous '
+                'or 1-dimensional.'
+            )
+
+    _map_array(input_arr, out, input_vals, output_vals)
+    return out.reshape(orig_shape)
+
+
+class ArrayMap:
```

Comment:
I may be missed something, but is the definition of the `ArrayMap` object necessary?

Reply:
@rfezzani I think the docstring does a decent job explaining things? It is a way to preserve the array indexing remapping API without creating enormous arrays.

Reply:
OK, my question then was not clear, sorry. I reformulate: can't we use a function here instead of the class definition?

Reply:
The class implements the `__array__` method for backward compatibility as `relabel_sequential` returns the forward and backward transformations in the form of ndarrays that are used as lookup tables. 
If we just returned simple functions for forward and backward transformations existing code that relies on arrays being returned would no longer work.
Basically we need the function to be triggered when `[ ]` indexing is used.

Reply:
OK, I was in fact missing something =), but I think that this approach is not addressing the problem of large memory allocation: it is simply postponed to the implementation of  the `__array__` method!

Reply:
@rfezzani no, the class allows relabeling with `[]` using Cython. It never needs to instantiate the arrays.

Reply:
I am confused here. If the `__array__`  method definition is required, the `MemoryError` pointed out in the PR description is not addressed...

Reply:
You are correct, one can still trigger a `MemoryError` if one wants to instantiate the lookup-table as a numpy array. However, instantiating the lookup-table is not required to access the functionality of relabeling. The same functionality is provided (without memory issues) using the `__getitem__` and `__call__` interfaces.

Example:
```
labels =  np.array([1, 1e16], dtype=np.int64)

# 1. call relabel_sequential
relab, fw, inv = relabel_sequential(labels) # does not allocate tons of memory

# 2. use the returned forward transform to relabel the array again using fancy indexing:
_relab = fw[labels] # does not allocate tons of memory ... is handled by __getitem__

# 3. use the returned forward transform to relabel the array again using call interdace:
_relab = fw(labels) # does not allocate tons of memory ... is handled by __call__

# 4. turn the returned forward transform into a numpy array 
fw_lut = np.narray(fw) # this will now try and allocate a huge array !
```

Use case `4.` is the only one where we still are likely to get `MemoryErrors` if we have large values in the input array. So in that sense one can still potentially hit this problem in the current code. However, this functionality is exclusively for backwards compatibility, in case anyone has code that used the returned transformations as an array (and not just for fancy indexing) for whatever reason. 
We provide the functionality for relabeling another array using the `fw` or `inv` transforms using methods `2.,3.`.

I agree that this is not immediately obvious. To address this I have the following suggestions:
1. spell out very clearly in the docstring that `np.array(fw)` is not recommended, or
1. raise a warning in `__array__`  that states that this is not recommended and for backwards compatibility only, or
1. remove the `__array__` method altogether, which may be a breaking change but will likely not affect a lot of use cases.